src/traefik.py

```diff
@@ -0,0 +1,574 @@
+RECV_CA_TEMPLATE = Template(f"{CA_CERTS_PATH}/receive-ca-cert-$rel_id-ca.crt")
+BIN_PATH = "/usr/bin/traefik"
+
+
+class RoutingMode(enum.Enum):
+    """Routing mode."""
+
+    path = "path"
+    subdomain = "subdomain"
+
+
+class TraefikError(RuntimeError):
+    """Base class for errors raised by this module."""
+
+
+class ContainerNotReadyError(TraefikError):
+    """Raised if the caller incorrectly assumes that the traefik container is ready."""
+
+
+class Traefik:
+    """Traefik workload representation."""
+
+    diagnostics_port = 8082  # Prometheus metrics, healthcheck/ping
+    _port = 80
+    _tls_port = 443
+    log_path = "/var/log/traefik.log"
+    _layer_name = "traefik"
+
+    def __init__(
+        self,
+        service_name: str,
+        container: Container,
+        routing_mode: RoutingMode,
+        tls_enabled: bool,
+        tcp_entrypoints: Dict[str, int],
+    ):
+        self._service_name = service_name
+        self._container = container
+        self._tcp_entrypoints = tcp_entrypoints
+        self._routing_mode = routing_mode
+        self._tls_enabled = tls_enabled
+
+    @property
+    def _hostname(self) -> str:
+        return socket.getfqdn()
+
+    @property
+    def service_ports(self) -> List[ServicePort]:
+        """Kubernetes service ports to be opened for this workload."""
+        web = ServicePort(self._port, name=f"{self._service_name}")
+        websecure = ServicePort(self._tls_port, name=f"{self._service_name}-tls")
+        return [web, websecure] + [
+            [ServicePort(int(port), name=name) for name, port in self._tcp_entrypoints.items()]
+        ]
+
+    @property
+    def scrape_jobs(self) -> list:
+        """List of static configs for prometheus scrape."""
+        return [
+            {
+                "static_configs": [{"targets": [f"{self._hostname}:{self.diagnostics_port}"]}],
+            }
+        ]
+
+    def _config_tls(self) -> str:
+        """Generate tls config yaml for traefik."""
+        config = yaml.safe_dump(
+            {
+                "tls": {
+                    "certificates": [
+                        {
+                            "certFile": SERVER_CERT_PATH,
+                            "keyFile": SERVER_KEY_PATH,
+                        }
+                    ],
+                    "stores": {
+                        "default": {
+                            # When the external hostname is a bare IP, traefik cannot match a domain,
+                            # so we must set the default cert for the TLS handshake to succeed.
+                            "defaultCertificate": {
+                                "certFile": SERVER_CERT_PATH,
+                                "keyFile": SERVER_KEY_PATH,
+                            },
+                        },
+                    },
+                }
+            }
+        )
+        self._container.push(DYNAMIC_CERTS_PATH, config, make_dirs=True)
```

Comment:
Given the push here, perhaps we should name the method more clearly?
"update tls config" or "push tls config"?